crates/matrix-sdk-common/src/deserialized_responses.rs

```diff
@@ -302,9 +307,59 @@ pub struct EncryptionInfo {
     /// Callers that persist this should mark the state as dirty when a device
     /// change is received down the sync.
     pub verification_state: VerificationState,
-    /// The Megolm session ID that was used to encrypt this event, or None if
-    /// this info was stored before we collected this data.
-    pub session_id: Option<String>,
+}
+
+impl EncryptionInfo {
+    /// Helper to get the megolm session id used to encrypt.
+    pub fn session_id(&self) -> Option<&str> {
+        let AlgorithmInfo::MegolmV1AesSha2 { session_id, .. } = &self.algorithm_info;
+        session_id.as_deref()
+    }
+}
+
+impl<'de> Deserialize<'de> for EncryptionInfo {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        // Deserialize as a generic JSON value
+        let value = serde_json::Value::deserialize(deserializer)?;
+
+        // Backwards compatibility:  Capture session_id at root if exists, in legacy
+        // EncryptionInfo the session_id was not in AlgorithmInfo
+        let binding = value.clone();
+        let s_id =
+            binding.as_object().and_then(|obj| obj.get("session_id")).and_then(|s| s.as_str());
+
+        // Now deserialize normally without legacy fields
+        #[derive(Deserialize)]
+        struct Helper {
+            pub sender: OwnedUserId,
+            pub sender_device: Option<OwnedDeviceId>,
+            pub algorithm_info: AlgorithmInfo,
+            pub verification_state: VerificationState,
+        }
```

Comment:
Can't this just be:

```rust
struct Helper {
    sender: OwnedUserId,
    sender_device: Option<OwnedDeviceId>,
    algorithm_info: AlgorithmInfo,
    verification_state: VerificationState,
    #[serde(rename = "session_id"]
    old_session_id,
}
```

The helper being the same as the real struct might confuse people. This also avoids two deserialization steps and the dependence on `serde_json::Value`.

Then you can merge the session IDs using:

```rust
let session_id = session_id.or(old_session_id);
```

It might make sense to destruct the helper, so any new additions to the struct don't get forgotten:

```rust
let Helper { sender, sender_device, algorithm_info, verification_state, session_id} = Helper::deserialize(deserializer)?;
```

Reply:
Thanks it is a lot better. see here https://github.com/matrix-org/matrix-rust-sdk/pull/4981/commits/69fe1db925a0ad77fed7467783fe98251bad37b0

We might write some sort of cheatsheet with all the strategies we use migrate to with serde (via a Helper, custom deserialize, field attributes, serde from/try_from, untagged...). I'll try to come with a first version